Add new onTestStarted and onTestFinished event listeners to the browser.test namespace

kiaraarose opened the pull request:
#### 1a5e68c92a29757c9a63b716382ce2c54b5839c3
<pre>
Add new onTestStarted and onTestFinished event listeners to the browser.test namespace
<a href="https://bugs.webkit.org/show_bug.cgi?id=294398">https://bugs.webkit.org/show_bug.cgi?id=294398</a>
<a href="https://rdar.apple.com/152970507">rdar://152970507</a>

Reviewed by Timothy Hatcher.

Instead of using the browser.test.onMessage event listener for bidirectional communication, create
new event listeners specially for reporting when a test has started and finished.

* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h:
(WebKit::toAPIString):
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
(-[WKWebExtensionContext _sendTestMessage:withArgument:]):
(-[WKWebExtensionContext _sendTestStartedMessageWithArgument:]):
(-[WKWebExtensionContext _sendTestFinishedMessageWithArgument:]):

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContextPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
(WebKit::isTestEventListener):
(WebKit::WebExtensionContext::addListener):
(WebKit::WebExtensionContext::removeListener):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::sendTestMessage):
(WebKit::WebExtensionContext::addTestMessageToQueue):
(WebKit::WebExtensionContext::sendQueuedTestMessagesIfNeeded):
(WebKit::WebExtensionContext::hasTestEventListeners):
Update these methods to handle testOnStarted and testOnFinished messages.

(WebKit::WebExtensionContext::flushTestMessageQueueIfNeeded): Deleted. Renamed.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::hasTestMessageEventListeners): Deleted.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionAPITest::onTestStarted):
(WebKit::WebExtensionAPITest::onTestFinished):
Create new event listeners.

(WebKit::WebExtensionContextProxy::dispatchTestMessageEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITest.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
Pass the event type so we can know which event listener to call for the message.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPITest, TestStartedEvent)):
(TestWebKitAPI::TEST(WKWebExtensionAPITest, TestFinishedEvent)):
Add new tests.

* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager sendTestStartedMessageWithArgument:]):
(-[TestWebExtensionManager sendTestFinishedMessageWithArgument:]):

Canonical link: <a href="https://commits.webkit.org/296322@main">https://commits.webkit.org/296322@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/885c30bfb25c54cf16bc802670fdf7321513c035

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27859 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18279 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113407 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110159 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28559 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36410 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/82153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111144 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/22630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/97463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/62584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/22045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/15599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/58143 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/91996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/15655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116533 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35262 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35636 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/93739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/90974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/35865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/13625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17471 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35161 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40717 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34895 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/38252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36562 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6b3a2e87698f2cd0007d8393e05711a903c4c269. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46679

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h:
```diff
@@ -76,6 +76,8 @@ enum class WebExtensionEventListenerType : uint8_t {
     TabsOnReplaced,
     TabsOnUpdated,
     TestOnMessage,
+    TestOnTestStarted,
+    TestOnTestFinished,
```
You need to bump `currentBackgroundContentListenerStateVersion` when adding to this.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -49,6 +49,50 @@
+TEST(WKWebExtensionAPITest, TestStartedEvent)
+{
+    auto *backgroundScript = Util::constructScript(@[
+        @"browser.test.onTestStarted.addListener((message, data) => {",
+        @"  browser.test.assertEq(message, 'test-started', 'message should be')",
```
I think we should just send data, there is no need for `message` — the event is the message now. This likely would need removed from the RFC too.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -49,6 +49,50 @@
+TEST(WKWebExtensionAPITest, TestFinishedEvent)
+{
+    auto *backgroundScript = Util::constructScript(@[
+        @"browser.test.onTestFinished.addListener((message, data) => {",
+        @"  browser.test.assertEq(message, 'test-finished', 'message should be')",
+        @"  browser.test.assertEq(data?.testName, 'test', 'data.testName should be')",
```
Ditto on dropping `message`.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContextPrivate.h:
```diff
@@ -45,6 +45,20 @@ WK_HEADER_AUDIT_BEGIN(nullability, sendability)
 
+/*!
+ @abstract Sends a message to the JavaScript `browser.test.onTestStarted` API.
+ @discussion Allows code to trigger a `browser.test.onTestStarted` event, enabling bidirectional communication during testing.
```
```suggestion
 @discussion Allows code to trigger a `browser.test.onTestStarted` event during testing.
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContextPrivate.h:
```diff
@@ -45,6 +45,20 @@ WK_HEADER_AUDIT_BEGIN(nullability, sendability)
+
+/*!
+ @abstract Sends a message to the JavaScript `browser.test.onTestFinished` API.
+ @discussion Allows code to trigger a `browser.test.onTestFinished` event, enabling bidirectional communication during testing.
```
```suggestion
 @discussion Allows code to trigger a `browser.test.onTestFinished` event during testing.
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContextPrivate.h:
```diff
@@ -45,6 +45,20 @@ WK_HEADER_AUDIT_BEGIN(nullability, sendability)
 - (void)_sendTestMessage:(NSString *)message withArgument:(nullable id)argument;
 
+/*!
+ @abstract Sends a message to the JavaScript `browser.test.onTestStarted` API.
```
```suggestion
 @abstract Sends data to the JavaScript `browser.test.onTestStarted` API.
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContextPrivate.h:
```diff
@@ -45,6 +45,20 @@ WK_HEADER_AUDIT_BEGIN(nullability, sendability)
+- (void)_sendTestStartedMessageWithArgument:(nullable id)argument;
+
+/*!
+ @abstract Sends a message to the JavaScript `browser.test.onTestFinished` API.
```
```suggestion
 @abstract Sends data to the JavaScript `browser.test.onTestFinished` API.
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -5007,36 +5007,89 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
-        sendTestMessage(testMessage.message, testMessage.argument.get());
+    switch (type) {
+    case WebExtensionEventListenerType::TestOnMessage:
+        m_testMessageQueue.append({ message, argument });
```
I think this is fine. If it becomes an issue we can fix it then.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
```diff
@@ -848,7 +848,17 @@ - (void)_sendTestMessage:(NSString *)message withArgument:(id)argument
+
+- (void)_sendTestStartedMessageWithArgument:(id)argument
+{
+    self._protectedWebExtensionContext->sendTestMessage("test-started"_s, argument, WebKit::WebExtensionEventListenerType::TestOnTestStarted);
```
```suggestion
    self._protectedWebExtensionContext->sendTestMessage(nullString(), argument, WebKit::WebExtensionEventListenerType::TestOnTestStarted);
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
```diff
@@ -848,7 +848,17 @@ - (void)_sendTestMessage:(NSString *)message withArgument:(id)argument
+
+- (void)_sendTestStartedMessageWithArgument:(id)argument
+{
+    self._protectedWebExtensionContext->sendTestMessage("test-started"_s, argument, WebKit::WebExtensionEventListenerType::TestOnTestStarted);
+}
+
+- (void)_sendTestFinishedMessageWithArgument:(id)argument
+{
+    self._protectedWebExtensionContext->sendTestMessage("test-finished"_s, argument, WebKit::WebExtensionEventListenerType::TestOnTestFinished);
```
```suggestion
    self._protectedWebExtensionContext->sendTestMessage(nullString(), argument, WebKit::WebExtensionEventListenerType::TestOnTestStarted);
}

- (void)_sendTestFinishedMessageWithArgument:(id)argument
{
    self._protectedWebExtensionContext->sendTestMessage("test-finished"_s, argument, WebKit::WebExtensionEventListenerType::TestOnTestFinished);
```

elijahsawyers reviewed (changes requested):
Marking changes requested so that I remember to come back and take another look once the feedback has been addressed.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c3be5f10a354b17060464ad6d5dd4c713c60b747. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46679

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c3be5f10a354b17060464ad6d5dd4c713c60b747. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46679

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2da9c23f0c39b71ec4fd6948d93c18c5d8973607. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46679

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3204aad57654d160b32fb35cdb51d1ecc2e4614b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46679

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -5017,16 +5017,15 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
 
-    if (!hasTestMessageEventListeners()) {
-        m_testMessageQueue.append({ message, argument });
+    constexpr auto eventType = WebExtensionEventListenerType::TestOnTestStarted;
```
cause of the build failures. should be `TestOnMessage`

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/65a7e20f0b7aeda0699156d0df42eb0d3e3e4760. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46679

kiaraarose committed:
Add new onTestStarted and onTestFinished event listeners to the browser.test namespace
https://bugs.webkit.org/show_bug.cgi?id=294398
rdar://152970507

Reviewed by Timothy Hatcher.

Instead of using the browser.test.onMessage event listener for bidirectional communication, create
new event listeners specially for reporting when a test has started and finished.

* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h:
(WebKit::toAPIString):
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
(-[WKWebExtensionContext _sendTestMessage:withArgument:]):
(-[WKWebExtensionContext _sendTestStartedMessageWithArgument:]):
(-[WKWebExtensionContext _sendTestFinishedMessageWithArgument:]):

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContextPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
(WebKit::isTestEventListener):
(WebKit::WebExtensionContext::addListener):
(WebKit::WebExtensionContext::removeListener):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::sendTestMessage):
(WebKit::WebExtensionContext::addTestMessageToQueue):
(WebKit::WebExtensionContext::sendQueuedTestMessagesIfNeeded):
(WebKit::WebExtensionContext::hasTestEventListeners):
Update these methods to handle testOnStarted and testOnFinished messages.

(WebKit::WebExtensionContext::flushTestMessageQueueIfNeeded): Deleted. Renamed.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::hasTestMessageEventListeners): Deleted.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionAPITest::onTestStarted):
(WebKit::WebExtensionAPITest::onTestFinished):
Create new event listeners.

(WebKit::WebExtensionContextProxy::dispatchTestMessageEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITest.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
Pass the event type so we can know which event listener to call for the message.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPITest, TestStartedEvent)):
(TestWebKitAPI::TEST(WKWebExtensionAPITest, TestFinishedEvent)):
Add new tests.

* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager sendTestStartedMessageWithArgument:]):
(-[TestWebExtensionManager sendTestFinishedMessageWithArgument:]):

Canonical link: https://commits.webkit.org/296322@main
